network/consensus/consensus.go

```diff
@@ -0,0 +1,65 @@
+/*
+ *    Copyright 2018 Insolar
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package consensus
+
+import (
+	"context"
+	"errors"
+
+	"github.com/insolar/insolar/core"
+)
+
+//go:generate stringer -type=step
+type step int
+
+const (
+	stepUndefined step = iota + 1
+	stepExchageData
+	stepExchageHash
+)
+
+// Participant describes one consensus participant
+type Participant interface {
+	GetID() core.RecordRef
+}
+
+// DataProvider for data manipulation
+type DataProvider interface {
+	GetDataList() []interface{}
+	MergeDataList([]interface{}, error)
+}
+
+// Consensus interface provides method to make consensus between participants
+type Consensus interface {
+	// DoConsensus is sync method, it make all consensus steps and returns boolean result
+	// method should be executed in goroutine
+	DoConsensus(ctx context.Context, self Participant, allParticipants []*Participant) (bool, error)
+}
+
+// Communicator interface is used to exchange messages between participants
+type Communicator interface {
+	// ExchangeData used in first consensus step to exchange data between participants
+	ExchangeData(ctx context.Context, p Participant, data []byte) ([]byte, error)
```

Comment:
I think we should not make these interfaces fully abstract; we could use data structs and interfaces from core modules instead, making consensus modules separated from network layer only. We already have a RecordRef dependency, so it is natural to use other core modules too (especially keeping in mind that using interface{} is pain)